internal/auth/providers/azure_graph.go

```diff
@@ -0,0 +1,170 @@
+package providers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// AzureGroupCacheSize controls the size of the caches of AD group info
+const AzureGroupCacheSize = 1024
+
+// GraphService wraps calls to provider admin APIs
+type GraphService interface {
+	GetGroups(string) ([]string, error)
+}
+
+// AzureGraphService implements graph API calls for the Azure provider
+type AzureGraphService struct {
+	client               *http.Client
+	groupMembershipCache *lru.Cache
+	groupNameCache       *lru.Cache
+}
+
+// NewAzureGraphService creates a new graph service for getting groups
+func NewAzureGraphService(clientID string, clientSecret string, tokenURL string) *AzureGraphService {
+	clientConfig := &clientcredentials.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		TokenURL:     tokenURL,
+		Scopes: []string{
+			"https://graph.microsoft.com/.default",
+		},
+	}
+	ctx := context.Background()
+	client := clientConfig.Client(ctx)
+	memberCache, err := lru.New(AzureGroupCacheSize)
+	if err != nil {
+		panic(err) // Should only happen if AzureGroupCacheSize is a negative number
+	}
+	nameCache, err := lru.New(AzureGroupCacheSize)
+	if err != nil {
+		panic(err) // Should only happen if AzureGroupCacheSize is a negative number
+	}
+	return &AzureGraphService{
+		client:               client,
+		groupMembershipCache: memberCache,
+		groupNameCache:       nameCache,
+	}
+}
+
+// GetGroups lists groups user belongs to.
+func (gs *AzureGraphService) GetGroups(email string) ([]string, error) {
+	if gs.client == nil {
+		return []string{}, errors.New("oauth client must be configured")
+	}
+	if email == "" {
+		return []string{}, errors.New("missing email")
+	}
+
+	var wg sync.WaitGroup
+	var mtx sync.Mutex
```

Comment:
nit: We've been using the name `mux` as an idiomatic spelling for a mutex variable elsewhere in the codebase - Would you mind using that spelling, for consistency?

Reply:
Sure, totally reasonable.